debian/bwbar.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.35.
.TH BWBAR "8" "Apr. 18, 2006" "bwbar 1.2.3"
.SH NAME
bwbar \- generates text and graphical readout of current bandwidth use. 
.SH SYNOPSIS
.B bwbar
[\fB\-i|-o\fR] [\fB\-g|-G|-k|-K|-m|-M\fR] [\fIOPTION\fR]... \fIINTERFACE\fR \fIMAX_MBPS\fR
.SH DESCRIPTION
Generates text and graphical readout of current bandwidth use.
.SH OPTIONS
.TP
\fB\-b\fR, \fB\-\-border\fR=\fIN\fR 
Set border width of the graphical bar to N (default 1 pixel).
.TP
\fB\-d\fR, \fB\-\-directory\fR=\fIN\fR 
Output directory is N.
.TP
\fB\-f\fR, \fB\-\-text\-file\fR=\fIN\fR
Set N as name of the text output file (default ubar.txt).
.TP
\fB\-h\fR, \fB\-\-help\fR
Display usage.
.TP
\fB\-p\fR, \fB\-\-png\-file\fR=\fIN\fR
Set N as name of the graphical bar file (default ubar.png).
.TP
\fB\-t\fR, \fB\-\-interval\fR=\fIN\fR
The poll interval in seconds is N (default 15 seconds).
.TP
\fB\-x\fR, \fB\-\-width\fR=\fIN\fR
Set the width of the graphical bar to N (default 600 pixels).
.TP
\fB\-y\fR, \fB\-\-height\fR=\fIN\fR
Set the height of the graphical bar to N (default 4 pixels).
.SH FLAGS
.TP
\fB\-i\fR, \fB\-\-input\fR
Measure input bandwidth.
.TP
\fB\-g\fR, \fB\-\-GBps\fR
Bandwidth is measured in GB/s.
.TP
\fB\-G\fR, \fB\-\-Gbps\fR
Bandwidth is measured in Gbit/s.
.TP
\fB\-k\fR, \fB\-\-kbps\fR
Bandwidth is measured in kbit/s.
.TP
\fB\-K\fR, \fB\-\-KBps\fR
Bandwidth is measured in KB/s.
.TP
\fB\-m\fR, \fB\-\-MBps\fR
Bandwidth is measured in MB/s.
.TP
\fB\-M\fR, \fB\-\-Mbps\fR
Bandwidth is measured in Mbit/s.
.TP
\fB\-o\fR, \fB\-\-output\fR
Measure output bandwidth.
.PP
Default values are \fB\-M\fR and \fB\-o\fR.
.SH AUTHOR
Written by Nacho Barrientos Arias <chipi@criptonita.com>.
.SH BUGS
Report bugs to <hpa@zytor.com>.
.SH COPYRIGHT
Copyright \(co 1999-2001 H. Peter Anvin - All Rights Reserved
.br
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, Inc.
.SH "SEE ALSO"
You can get all the available information for
.B bwbar
in:
.IP
.B http://www.kernel.org/pub/software/web/bwbar/
.PP


debian/changelog
bwbar (1.2.3-2) unstable; urgency=low

  * Fix FTBFS with libpng 1.5,
    thanks to Nobuhiro Iwamatsu<iwamatsu@nigauri.org>
    (Closes: #635742)
  * Update dh compat to 7
  * Bump standards version

 -- Julien Danjou <acid@debian.org>  Thu, 28 Jul 2011 16:13:14 +0200

bwbar (1.2.3-1) unstable; urgency=low

  * New upstream release
  * Bump standards version
  * Add LSB section to init script
  * Enhance init script

 -- Julien Danjou <acid@debian.org>  Mon, 11 Sep 2006 14:05:37 +0200

bwbar (1.2.2-4) unstable; urgency=low

  * New maintainer
  * Bump standards version
  * Use debhelper 4
  * Remove http://www.kernel.org from description since this is no more
    accurate
  * Update README.Debian
  * Rework patches
    - Handle new options correctly (Closes: #313231)
  * Add manpage (Closes: #129668)
  * Fix {pre,post}{rm,inst} scripts to use invoke-rc.d

 -- Julien Danjou <acid@debian.org>  Sun, 23 Apr 2006 13:11:09 +0200

bwbar (1.2.2-3.1) unstable; urgency=low

  * NMU
  * Fix a wrong setting in rules, causing FTBFS (Closes: #358940)

 -- Julien Danjou <acid@debian.org>  Fri, 21 Apr 2006 18:03:11 +0200

bwbar (1.2.2-3) unstable; urgency=low

  * Applied patch from Adrian Reber that fixed the short flag (d) for
    directory. (Closes: #260999)

 -- Scott M. Dier <sdier@debian.org>  Sat, 24 Jul 2004 00:20:02 -0500

bwbar (1.2.2-2) unstable; urgency=low

  * Added Build-Depends on cdbs.

 -- Scott M. Dier <sdier@debian.org>  Tue,  8 Jun 2004 07:36:25 -0500

bwbar (1.2.2-1) unstable; urgency=low

  * New upstream release (Closes: #250730)
  * Converted to cdbs.
  * Added patch to support {K,M,G}B/s. (Closes: #249852)

 -- Scott M. Dier <sdier@debian.org>  Tue,  1 Jun 2004 15:01:34 -0500

bwbar (1.2-7) unstable; urgency=low

  * Fixed fscanf call in bwbar.c for longer interface names.
    (Closes: #246054)
  * Fixed README for inaccurate configuration information. (Closes: #229162)

 -- Scott M. Dier <sdier@debian.org>  Wed, 12 May 2004 01:01:48 -0500

bwbar (1.2-6) unstable; urgency=low

  * Fixed init.d message about being unconfigured. (Closes: #199202)
  * Recompiled against libpng3. (Closes: #180243)

 -- Scott M. Dier <sdier@debian.org>  Wed, 10 Dec 2003 20:44:05 -0600

bwbar (1.2-5) unstable; urgency=low

  * Fixed that I didn't reperm the init.d script, this causes issues
  because diffs don't keep permissions.  Closes: #122310

 -- Scott M. Dier <sdier@debian.org>  Tue,  4 Dec 2001 01:08:37 -0600

bwbar (1.2-4) unstable; urgency=low

  * Didn't copy a line right in the init.d script, fixed. Closes: #119536

 -- Scott M. Dier <sdier@debian.org>  Wed, 14 Nov 2001 14:31:34 -0600

bwbar (1.2-3) unstable; urgency=low

  * Fixed build-deps, added libpng2-dev.  closes: #114375

 -- Scott M. Dier <sdier@debian.org>  Sat,  6 Oct 2001 20:27:46 -0500

bwbar (1.2-2) unstable; urgency=low

  * Added --directory function to daemon.

 -- Scott M. Dier <sdier@debian.org>  Mon,  1 Oct 2001 02:00:52 -0500

bwbar (1.2-1) unstable; urgency=low

  * Initial Release.

 -- Scott M. Dier <sdier@debian.org>  Sun, 30 Sep 2001 23:35:14 -0500

Local variables:
mode: debian-changelog
End:


bwbar.c
/* $Id: bwbar.c,v 1.7 2006/08/26 16:31:26 hpa Exp $ */
/* ----------------------------------------------------------------------- *
 *   
 *   Copyright 1999-2006 H. Peter Anvin - All Rights Reserved
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, Inc., 675 Mass Ave, Cambridge MA 02139,
 *   USA; either version 2 of the License, or (at your option) any later
 *   version; incorporated herein by reference.
 *
 * ----------------------------------------------------------------------- */

/*
 * bwbar.c
 *
 * Bandwidth monitor using /proc/net/dev which produces output which can
 * be displayed on a web page.
 *
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <errno.h>
#include <time.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>
#include <alloca.h>
#include <getopt.h>
#include <png.h>
#include <zlib.h>

void skipline(FILE *f)
{
  int ch;
  do {
    ch = getc(f);
  } while ( ch != '\n' && ch != EOF );
}

int write_bar_graph(FILE *f, double perc, int width, int height, int border, int spacers)
{
  static png_color palette[4] =
  {
    {  0,   0,   0},		/* RGB color for border */
    {255, 255, 255},		/* RGB color for separators */
    {214, 136, 131},		/* RGB color for unfilled part of graph */
    {204, 204, 227},		/* RGB color for filled part of graph */
  };
  int bwidth, bheight;		/* Width and height including border */
  double frac;
  int spacer, fspacer;
  int x, y;
  int color;
  png_structp png_ptr = NULL;
  png_infop info_ptr  = NULL;
  png_byte *border_row;
  png_byte *bar_row;
  png_byte *bp;
  int status = 1;		/* Assume failure */

  bwidth = width + border*2;
  bheight = height + border*2;

  border_row   = alloca(sizeof(png_byte) * bwidth);
  bar_row      = alloca(sizeof(png_byte) * bwidth);

  if ( !border_row || !bar_row )
    goto barf;

  png_ptr = png_create_write_struct(PNG_LIBPNG_VER_STRING,
				    NULL, NULL, NULL);

  if ( !png_ptr )
    goto barf;

  info_ptr = png_create_info_struct(png_ptr);
  if ( !info_ptr )
    goto barf;

  if ( setjmp(png_jmpbuf((png_ptr))) ) {
    status = 1;
    goto barf;			/* libpng abort */
  }

  png_init_io(png_ptr, f);

  png_set_compression_level(png_ptr, Z_BEST_COMPRESSION);

  png_set_IHDR(png_ptr, info_ptr, bwidth, bheight,
	       2, PNG_COLOR_TYPE_PALETTE, PNG_INTERLACE_NONE,
	       PNG_COMPRESSION_TYPE_DEFAULT,
	       PNG_FILTER_TYPE_DEFAULT);

  png_set_PLTE(png_ptr, info_ptr, palette, 4);

  png_write_info(png_ptr, info_ptr);

  png_set_packing(png_ptr);

  /* Now we actually produce the data */

  /* Generate the border row */
  memset(border_row, 0, bwidth);

  /* Generate the bar row */
  bp = bar_row;
  for ( x = 0 ; x < border ; x++ )
    *bp++ = 0;			/* Border color */

  spacer = 0;
  for ( x = 1 ; x <= width ; x++ ) {
    frac = (double)x/(double)width;
    fspacer = (int)(spacers*frac);
    if ( x < width && spacer < fspacer ) {
      color = 1;		/* Spacer color */
    } else if ( perc < frac ) {
      color = 3;		/* Bar color */
    } else {
      color = 2;		/* Background color */
    }
    *bp++ = color;
    spacer = fspacer;
  }
  
  for ( x = 0 ; x < border ; x++ )
    *bp++ = 0;			/* Border color */

  /* Write output */
  for ( y = 0 ; y < border ; y++ )
    png_write_row(png_ptr, border_row);
  for ( y = 0 ; y < height ; y++ )
    png_write_row(png_ptr, bar_row);
  for ( y = 0 ; y < border ; y++ )
    png_write_row(png_ptr, border_row);

  /* Done writing */
  png_write_end(png_ptr, info_ptr);
  status = 0;			/* Success! */

 barf:				/* Jump here on fatal error */
  png_destroy_write_struct(&png_ptr, &info_ptr);
  return status;
}

const struct option longopts[] = {
  { "input",        0, 0, 'i' },
  { "output",       0, 0, 'o' },
  { "text-file",    1, 0, 'f' },
  { "png-file",     1, 0, 'p' },
  { "interval",     1, 0, 't' },
  { "width",        1, 0, 'x' },
  { "height",       1, 0, 'y' },
  { "border",       1, 0, 'b' },
  { "kbps",         0, 0, 'k' },
  { "Mbps",         0, 0, 'M' },
  { "Gbps",         0, 0, 'G' },
  { "help",         0, 0, 'h' },
  { 0, 0, 0, 0 }
};

const char *program;

void usage(int err)
{
  fprintf(stderr,
	  "Usage: %s [options] interface max_mbps\n"
	  "Options: (defaults in parenthesis)\n"
	  "   --input               -i   Measure input bandwidth\n"
	  "   --output              -o   Measure output bandwidth (default)\n"
	  "   --text-file <file>    -f   The name of the text output file (ubar.txt)\n"
	  "   --png-file <file>     -p   The name of the graphical bar file (ubar.png)\n"
	  "   --interval <seconds>  -t   The poll interval in seconds (15)\n"
	  "   --width <pixels>      -x   Width of the graphical bar (600)\n"
	  "   --height <pixels>     -y   Height of the graphical bar (4)\n"
	  "   --border <pixels>     -b   Border width of the graphical bar (1)\n"
	  "   --kbps                -k   Bandwidth is measured in kbit/s\n"
	  "   --Mbps                -M   Bandwidth is measured in Mbit/s (default)\n"
	  "   --Gbps                -G   Bandwidth is measured in Gbit/s\n"
	  "   --help                -h   Display this text\n",
	  program);
  exit(err);
}

int main(int argc, char *argv[])
{
  FILE *pnd;
  char *interface;
  struct ifinfo {
    char name[8];
    unsigned int r_bytes, r_pkt, r_err, r_drop, r_fifo, r_frame;
    unsigned int r_compr, r_mcast;
    unsigned int x_bytes, x_pkt, x_err, x_drop, x_fifo, x_coll;
    unsigned int x_carrier, x_compr;
  } ifc;
  unsigned long long bin, bout, lbin, lbout;
  int first;
  struct timeval t_now, t_last;
  double maxbandwidth;
  double bwin, bwout, bwmeasure, timedelta;
  int opt;
  char *t_tmp, *g_tmp;
  /* Options */
  int measure_input = 0;	/* Input instead of output */
  char *text_file = "ubar.txt";	/* Text filename */
  char *graphics_file = "ubar.png"; /* Graphics filename */
  char *unit_name = "Mbit/s";	/* Unit name */
  double unit = 1.0e+6;		/* Unit multiplier */
  int interval = 15;		/* Interval between measurements (s) */
  int width = 600;		/* Bar width */
  int height = 4;		/* Bar height */
  int border = 1;		/* Bar border */

  program = argv[0];

  while ( (opt = getopt_long(argc, argv, "iof:p:t:x:y:b:kMGh", longopts, NULL)) != -1 ) {
    switch ( opt ) {
    case 'i':
      measure_input = 1;
      break;
    case 'o':
      measure_input = 0;
      break;
    case 'f':
      text_file = optarg;
      break;
    case 'p':
      graphics_file = optarg;
      break;
    case 't':
      interval = atoi(optarg);
      break;
    case 'x':
      width = atoi(optarg);
      break;
    case 'y':
      height = atoi(optarg);
      break;
    case 'b':
      border = atoi(optarg);
      break;
    case 'k':
      unit = 1.0e+3;
      unit_name = "kbit/s";
      break;
    case 'M':
      unit = 1.0e+6;
      unit_name = "Mbit/s";
      break;
    case 'G':
      unit = 1.0e+9;
      unit_name = "Gbit/s";
      break;
    case 'h':
      usage(0);
      break;
    default:
      usage(1);
      break;
    }
  }

  if ( argc-optind != 2 )
    usage(1);

  t_tmp = malloc(strlen(text_file) + 5);
  g_tmp = malloc(strlen(graphics_file) + 5);
  if ( !t_tmp || !g_tmp ) {
    perror(program);
    exit(1);
  }
  sprintf(t_tmp, "%s.tmp", text_file);
  sprintf(g_tmp, "%s.tmp", graphics_file);

  interface = argv[optind];
  maxbandwidth = atof(argv[optind+1]) * unit;

  first = 1;
  lbin = 0; lbout = 0;

  gettimeofday(&t_last, NULL);
  
  while ( 1 ) {

    /**** Begin code that obtains bandwidth data ****/

    gettimeofday(&t_now, NULL);
    pnd = fopen("/proc/net/dev", "r");
    if ( !pnd ) {
      fprintf(stderr, "%s: /proc/net/dev: %s", argv[0], strerror(errno));
      exit(1);
    }

    /* Skip header */
    skipline(pnd);
    skipline(pnd);

    /* Get interface info */
    do {
      if ( fscanf(pnd, " %[^:]:%u %u %u %u %u %u %u %u %u %u %u %u %u %u %u %u",
		  ifc.name,
		  &ifc.r_bytes, &ifc.r_pkt, &ifc.r_err, &ifc.r_drop,
		  &ifc.r_fifo, &ifc.r_frame, &ifc.r_compr, &ifc.r_mcast,
		  &ifc.x_bytes, &ifc.x_pkt, &ifc.x_err, &ifc.x_drop,
		  &ifc.x_fifo, &ifc.x_coll, &ifc.x_carrier, &ifc.x_compr)
	   != 17 ) {
	exit(200);
      }
      skipline(pnd);
    } while ( strcmp(ifc.name, interface) );

    bin  = ifc.r_bytes + (lbin & ~0xffffffffULL);
    bout = ifc.x_bytes + (lbout & ~0xffffffffULL);

    if ( bin < lbin )
      bin += (1ULL << 32);
    if ( bout < lbout )
      bout += (1ULL << 32);

    /**** Begin code that generates output ****/

    if ( !first ) {
      FILE *ubar, *pngmaker;

      timedelta = (double)(t_now.tv_sec - t_last.tv_sec) + 
	(t_now.tv_usec - t_last.tv_usec)/1.0e+6;
      bwin = (bin-lbin)*8/timedelta;
      bwout = (bout-lbout)*8/timedelta;

      bwmeasure = measure_input ? bwin : bwout;
      
      ubar = fopen(t_tmp, "w");
      pngmaker = fopen(g_tmp, "w");
      if ( ubar ) {
	fprintf(ubar, "Current bandwidth utilization %6.2f %s\n", bwmeasure/unit, unit_name);
	fclose(ubar);
	rename(t_tmp, text_file);
      }
      if ( pngmaker ) {
	write_bar_graph(pngmaker, bwmeasure/maxbandwidth, width, height, border, 5);
	fclose(pngmaker);
	rename(g_tmp, graphics_file);
      }
    } else {
      first = 0;
    }

    lbin = bin;  lbout = bout;  t_last = t_now;

    fclose(pnd);

    sleep(interval);
  }
}
